Client/game_sa/CPoolsSA.cpp

```diff
@@ -871,28 +872,34 @@ CVehicle* CPoolsSA::AddTrain(CClientVehicle* pClientVehicle, CVector* vecPositio
 
     // Find closest track node
     float fRailDistance;
-    int   iNodeId = pGame->GetWorld()->FindClosestRailTrackNode(*vecPosition, ucTrackId, fRailDistance);
-    int   iDesiredTrackId = ucTrackId;
+    auto pTrainTrackManager = g_pCore->GetGame()->GetTrainTrackManager();
 
-    DWORD dwFunc = FUNC_CTrain_CreateMissionTrain;
-    _asm
-    {
-        push    0            // place as close to point as possible (rather than at node)? (maybe) (actually seems to have an effect on the speed, so changed from
-                             // 1 to 0)
-                             push    iDesiredTrackId            // track ID
-                             push    iNodeId            // node to start at (-1 for closest node)
-                             lea     ecx, pTrainEnd
-                             push    ecx            // end of train
-                             lea     ecx, pTrainBeginning
-                             push    ecx            // begining of train
-                             push    0            // train type (always use 0 as thats where we're writing to)
-                             push    bDirection            // direction
-                             push    fZ            // z
-                             push    fY            // y
-                             push    fX            // x
-                             call    dwFunc
-                             add     esp, 0x28
-    }
+    auto nodeIndex = pGame->GetWorld()->FindClosestRailTrackNode(*vecPosition, pTrainTrack, fRailDistance); // TODO
+    uint desiredTrackId = pTrainTrack->GetIndex();
+
+    // TODO(qaisjp): this actually returns a CTrainSA, but we don't have that class!
+    typedef CVehicleSA*(__cdecl* CTrain_CreateMissionTrain)(float fX, float fY, float fZ, bool bDirection, int trainType, CVehicleSAInterface** pTrainBeginning,
+                                           CVehicleSAInterface** pTrainEnd, unsigned int nodeIndex, unsigned int desiredTrackId, char unk);
+    CTrain_CreateMissionTrain createMissionTrain = (CTrain_CreateMissionTrain)(FUNC_CTrain_CreateMissionTrain);
+
+    createMissionTrain(
+        fX, fY, fZ, bDirection,
+
+        // NOTE(qaisjp): I have no idea what the below comment means
+        // train type (always use 0 as thats where we're writing to)
+        0,
+
+        &pTrainBeginning, &pTrainEnd,
+
+        // Node to start at (-1 for closest node)
+        nodeIndex,
+
+        // track ID
+        desiredTrackId,
+
+        // NOTE(qaisjp): I have no idea what the below comment is talking about.
+        // place as close to point as possible (rather than at node)? (maybe) (actually seems to have an effect on the speed, so changed from 1 to 0)
+        0);
 
```

Comment:
> There is one problem: you can't set the train position of streamed out trains. Try placing two trains (`t1` and `t21`) on the same track, near each other (trainPositions `p1` and `p2`). Then disconnect, and use your server console to place them at the exact same position (set `p1 = p2`). Then reconnect (or restart the game and reconnect) and you'll find that the trains have moved.



TODO: we could set the train position here? otherwise the client's train will not match what the server requested it to be. there might be a better time for setting the train position than here, though, such as when the train streams in